### ISSUE TITLE
[$500] LHN - Public-room-regression does not appear in LHN for the same user on Web

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#29777]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.85-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
Precondition:
Login with same user on Android app and Web
1. Navigate to https://staging.new.expensify.com/r/5408450846930023
on Android app to open Public room
2. Go to Web app with same user
3. Check LHN

## Expected Result:
Public room should appear in LHN on Web

## Actual Result:
Public-room-regression does not appear in LHN for the same user on Web even when page reloads

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/78819774/17121254-1c8b-41dd-a44f-5069ad82dcd6

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0167ca73cae5d84314</li>
        <li>Upwork Job ID: 1714284800012386304</li>
        <li>Last Price Increase: 2023-11-07</li>
    </ul>
</details>

**Comment by laurenreidexpensify** [comment]
waiting for proposals

**Comment by akamefi202** [comment]
I think this is a back-end issue.
If the user opens a public room in mobile device, it calls OpenReport API.
The back-end uses Pusher package to send data to clients.
But OpenReport API doesn't send data of newly opened report to client devices of the user.
So the public room won't be added to LHN in web browser of the user.

Instead, AddComment API sends data of newly opened report to client devices.
As a result, the user can see the public room in LHN after he sent a message in the public room of mobile app.

Another example is when starting new chat between two users.
User B is able to see the new chat report after user A created the chat report & sent a message.

So to fix this issue, the back-end should send data of new public room report to client devices of the user while OpenReport API processing.

**Comment by laurenreidexpensify** [comment]
@mountiny @luacmartins can you take a look [here](https://github.com/Expensify/App/issues/29777#issuecomment-1777190818) - does this feel like `internal` to you?

**Comment by luacmartins** [comment]
I was not able to reproduce this issue. I saw the room in all devices I tried. 

>even when page reloads
This part is weird, because even if the user missed the Pusher update they should still get the report data in the ReconnectApp call.

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
1) Login with user A on the web and user A on another device
2) Navigate to https://staging.new.expensify.com/r/5408450846930023 on the web
3) Public-room-regression opens the web but does not appear on the other device.

### What is the root cause of that problem?
Pusher notifications are being handled for report actions only and not for reports.

### What changes do you think we should make in order to solve the problem?
Assuming both the devices are logged in as user A. When public channel like public-room-regression is opened on the web, the user logged in on the other device recieves pusher notification but the notification is not handled for reports only report actions.

The current implementation of triggerNotifications only handles notifications for reportactions
https://github.com/Expensify/App/blob/32603b659d37cec98a9d10e4cb7a50c1f51c77e4/src/libs/actions/User.js#L466

The triggerNotifications can be modified to handle report notifications.

```
function triggerNotifications(onyxUpdates) {
    _.each(onyxUpdates, (update) => {
        const shouldNotify = lodashGet(update, 'shouldNotify', true);
        if (!shouldNotify) {
            return;
        }

        if(update.key.includes('report_'))
        {
            const reportID = update.key.replace(ONYXKEYS.COLLECTION.REPORT, '');
            const report = ReportUtils.getReport(reportID);
            if (_.isEmpty(report) || reportID != '')
            {
                Report.openReport(reportID);
            }
        }
        else 
        {
            const reportID = update.key.replace(ONYXKEYS.COLLECTION.REPORT_ACTIONS, '');
            const reportActions = _.values(update.value);
    
            // eslint-disable-next-line rulesdir/no-negated-variables
            const notifiableActions = _.filter(reportActions, (action) => ReportActionsUtils.isNotifiableReportAction(action));
            _.each(notifiableActions, (action) => Report.showReportActionNotification(reportID, action));
        }

    });
}

```

### What alternative solutions did you explore? (Optional)

**Comment by sourcecodedeveloper** [comment]
@laurenreidexpensify proposal attached. Took some effort but I found out that the pusher notification can be handled on the front end.
@mountiny When you get some time can you help test the proposal.

**Comment by mountiny** [comment]
I feel like this is low value issue and if its not reliably reproducible then even more so

**Comment by sourcecodedeveloper** [comment]
@mountiny I can confirm the issue is reproducible on main, staging and also production. The android app needs to be restarted in order to get the public room visible. 
Video attached for main
[web_android_main.webm](https://github.com/Expensify/App/assets/141340971/65369423-2150-4bc3-bafe-88fa11cb4d2d)

**Comment by mountiny** [comment]
Why is this an issue only for public reports and not any thread or some other report you visit in one client and not in others

**Comment by sourcecodedeveloper** [comment]
The issue also happens when creating chat between two users by creating chat room
steps
Login with User A on web
1) Click on start chat
2) Create room, set workspace name and set visibility to public

Login with User B on web
Login with user B on android

1) From User B on web, navigate to the chatroom created by User A.
2) The room loads for User B on the web but the room does not load for User B on android.

@mountiny Better to ask that question from the Applause - Internal Team as they reported the issue in the first place.

**Comment by mountiny** [comment]
@ntdiary What do you think of the proposed solution from @sourcecodedeveloper did you test out this works fine?

**Comment by sourcecodedeveloper** [comment]
@ntdiary The backend pushes report data by pusher notification and not https. You only checked https. Check the video posted by QA carefully. The expectation is that after you opened the room on android using user A, the user A logged in on the web should also be able to view the room, check ending video posted by QA. 

Please test the proposal in your local environment. Update the triggerNotifications in User.js as mentioned in the proposal. Then follow the steps mentioned in the QA video. If you followed the steps correct, the pusher notification should be handled correctly by triggerNotifications.

**Comment by sourcecodedeveloper** [comment]
@ntdiary I am not challenging your skill set. Can you please test the proposal in your local environment. Then share your findings. Thanks.

**Comment by sourcecodedeveloper** [comment]
> The action steps in the OP are actually inconsistent with its video.
> 
>    Precondition:
>    Login with same user on Android app and Web
> 

@ntdiary The OP might have bought up separate issues. Need to confirm if push notifications change following preconditions. Assuming you are testing the proposal i.e. have modified the triggerNotifications in your local environment, please also test following preconditions. Then while testing based on preconditions you can also check for push notifications although the data is probably encrypted.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mountiny** [comment]
Seems like this could be internal PR indeed

**Comment by sourcecodedeveloper** [comment]
I just want to confirm the findings for the preconditions because the OP could have brought up separate issues.

**Comment by sourcecodedeveloper** [comment]
Still waiting on @ntdiary to test proposal based on preconditions and need to confirm if OP has brought up separate issue.
The issue can be tested by either following preconditions or by using deep-link. The deep-link approach has been tested, just need to test proposal based on preconditions.

@mountiny please do not close the issue or label internal yet, need to confirm if dealing with separate issues and waiting for findings based on preconditions.

**Comment by sourcecodedeveloper** [comment]
@ntdiary The [pusher](https://pusher.com/) library is independent of the backend and the pusher notifications are being received on the front end for the preconditions testing scenario and therefore the issue can be resolved on the front-end. 

> The backend will push report_5408450846930023 data in multiple scenarios, if we fetch the data again through Report.openReport(reportID) solely to fix this issue, it will result in many unnecessary requests from the client side. : )

the Report.openReport is only called when handling pusher events and not https and specifically when the event type is MULTIPLE_EVENTS, there would not many unnecessary requests.

You tested based on deep-linking due to inconsistency in the OP action steps and did not receive any push notifications -agreed but you also need to collaborate when requested and test the reported issue and proposal.

@mountiny I get the feeling that the C+ assigned to the issue is not collaborating in testing the proposal which is what you had asked for. If the C+ continues to make up excuses and refuses to test the reported issue and proposal perhaps the necessary thing to do is to assign another C+.

**Comment by sourcecodedeveloper** [comment]
> The root cause of this issue is that the backend did not push data to the frontend under appropriate conditions and timing.

@ntdiary  Again not backend issue and not the correct root cause. Under preconditions, the data is being sent to the front end by pusher event "multipleEvents". You can test based on preconditions and confirm data is encrypted and is being pushed by pusher channel.

```
{
    "event": "multipleEvents",
    "data": "{...}",
    "channel": "..."
}
```

> Pusher is just a message broker between the backend and frontend. The backend pushes messages to the frontend through it.

You seem to have some misunderstanding about Pusher, backend does not push data or messages to the frontend, the backend uses Api and sends response by https(XHR response) e.g. API.write('leaveroom') The pusher library can be used to subscribe to events or send events e.g. user leaving room.


> The report_xxx data received above will all be handled by triggerNotifications, and will call Report.openReport(reportID) for each one, which will cause many unnecessary openReport requests.

I have tested proposal based on your scenarios and triggerNotifications is only called if the report is empty or not available in the LHN. I can confirm there will not be many unnecessary openReport requests. The check if report is empty ensures the openReport is only called when the report is empty or undefined.

> The report data I'm referring to here does not specifically mean the report_xxx data, but rather all the data that can make this report show up in the LHN.

Yes that is what ReportUtils.getReport is used for to check if reports exists. If the report exists then ReportUtils.getReport would return the report.

> Furthermore, triggerNotifications is only used to display report action notifications on web/desktop, we don't need to add other logic there, and your proposal would also introduce potential issues (it assumes shouldNotify is true by default).

Your assumption about triggerNotifications only being used to display report on web/desktop is wrong. I would not propose solution if it was not cross-platform. triggerNotifications is cross-platform and applies to Android and IOS. I have tested the proposal and the triggerNotifications applies to Android and IOS. Please do not make baseless claims without actually testing the proposal in your local environment.

Again, I request that you test the proposal based on the preconditions and come back with your findings like you did for deep-linking. While testing under preconditions you can also verify that data is being pushed by pusher library (not backend). After you post your findings, I have no issue with collaboration.
 
@mountiny the proposal is cross-platform and not platform-specific i.e only for web/desktop. The debate is not whether the issue is back-end or internal but whether the scenarios preconditions and deep-linking are separate issues and another issue needs to be created for deep-linking.

**Comment by sourcecodedeveloper** [comment]
@mountiny
To summarize, the reported issue is not backend and proposal needs to be tested for loggedIn scenario and need to assign another C+ who actually tests the proposal before claiming triggerNotifications are web/desktop only.

**Comment by laurenreidexpensify** [comment]
Hi everyone, just catching up on where we're at in this issue. Reviewing the proposals so far and input from @ntdiary and @mountiny it looks like we haven't found a suitable proposal yet. 

@sourcecodedeveloper thank you for your proposals so far, but as per the C+ review we haven't yet found a suitable solution in those. It might be helpful at this point to put a pause on that direction, and see if anyone else in the community has an alternate perspective. Thanks

**Comment by sourcecodedeveloper** [comment]
@laurenreidexpensify the proposal and the reported issue for preloggedIn scenario has not been tested by the C+. If by alternate direction you mean resolving the issue on the backend, the issue is not backend. The solution is still on frontend.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week) are we ok to close this?

**Comment by laurenreidexpensify** [comment]
I am closing the issue based on the KI retesting results
